jupiterone/provider.go

```diff
@@ -4,65 +4,143 @@ import (
+// JupiterOneProvider contains the initialized API client to communicate with the JupiterOne API
+type JupiterOneProvider struct {
+	// version is set to the provider version on release, "dev" when the
+	// provider is built and ran locally, and "test" when running acceptance
+	// testing.
+	version string
+	Client  *client.JupiterOneClient
+}
+
+type JupiterOneProviderModel struct {
+	APIKey    basetypes.StringValue `tfsdk:"api_key"`
+	AccountID basetypes.StringValue `tfsdk:"account_id"`
+	Region    basetypes.StringValue `tfsdk:"region"`
+
+	// httpClient should _only_ be used for tests using go-vcr/cassettes
+	httpClient *http.Client
+}
+
+var _ provider.Provider = &JupiterOneProvider{}
+
+func New(version string) func() provider.Provider {
+	return func() provider.Provider {
+		return &JupiterOneProvider{
+			version: version,
+		}
 	}
 }
 
-// ProviderConfiguration contains the initialized API client to communicate with the JupiterOne API
-type ProviderConfiguration struct {
-	Client *client.JupiterOneClient
+// Configure implements provider.Provider
+func (p *JupiterOneProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	var data JupiterOneProviderModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if p.Client == nil {
+		// Check environment variables. Performing this as part of Configure is
+		// the current de-facto way of "merging" defaults:
+		// https://github.com/hashicorp/terraform-plugin-framework/issues/539#issuecomment-1334470425
+		var err error
+		p.Client, err = data.Client(ctx)
+
+		if err != nil {
+			resp.Diagnostics.AddError("failed to create JupiterOne client in provider configuration: %s", err.Error())
+			return
+		}
+		log.Println("[INFO] JupiterOne client successfully initialized")
+	} else {
+		log.Println("[INFO] Using already configured client")
+	}
+
+	resp.DataSourceData = p
+	resp.ResourceData = p
 }
 
-func providerConfigure(_ context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
-	log.Println("[INFO] JupiterOne client successfully initialized")
+// NewClient configures the J1 client itself from the provider model and
+// allows overrides from the env variables.
+//
+// TODO: For testing, will also accept an `http_client` set in the context for
+// for use with `cassettes` tests. This is a result of updating to the new
+// terraform-plugin-framework and needing to inject the client, while not
+// duplicating code or messing up the existing cassettes.
+func (m *JupiterOneProviderModel) Client(ctx context.Context) (*client.JupiterOneClient, error) {
+	apiKey := os.Getenv("JUPITERONE_API_KEY")
+	accountId := os.Getenv("JUPITERONE_ACCOUNT_ID")
+	region := os.Getenv("JUPITERONE_REGION")
+
+	if apiKey == "" {
+		apiKey = m.APIKey.ValueString()
+	}
+	if accountId == "" {
+		accountId = m.AccountID.ValueString()
+	}
+	if region == "" {
+		region = m.Region.ValueString()
+	}
```

Comment:
Yes, I've moved this to the provider's `Configure` function and made a separate `NewClientFromEnv` for use in the tests. I had put the logic in `(*ProviderModel) Client()` to avoid duplication, but I think having more explicit constructor for test clients in the `client` package is an improvement for this iteration. PTAL.